dojo/tools/checkmarx_osa/parser.py

```diff
@@ -0,0 +1,99 @@
+import json
+import logging
+from dojo.models import Finding
+from datetime import datetime
+
+logger = logging.getLogger(__name__)
+
+
+class CheckmarxOsaParser(object):
+
+    def get_scan_types(self):
+        return ["Checkmarx OSA"]
+
+    def get_label_for_scan_types(self, scan_type):
+        return scan_type  # no custom label for now
+
+    def get_description_for_scan_types(self, scan_type):
+        return "Checkmarx Open Source Analysis for dependencies (json). Generate with `jq -s . CxOSAVulnerabilities.json CxOSALibraries.json`"
+
+    def get_findings(self, filehandle, test):
+        tree = json.load(filehandle)
+        if len(tree) != 2:
+            logger.error("Bad format. Expected a list of two elements: CxOSALibraries.json and CxOSAVulnerabilities.json. Found %i elements", len(tree))
+            raise ValueError("Invalid format: bad structure")
+        libraries_dict = self.get_libraries(tree)
+        vulnerabilities = self.get_vunlerabilities(tree)
+        items = []
+        for item in vulnerabilities:
+            mandatory_vulnerability_fields = ['libraryId', 'state', 'severity']
+            mandatory_library_fields = ['name', 'version']
+            self.check_mandatory(item, mandatory_vulnerability_fields)
+            library = libraries_dict[item['libraryId']]
+            self.check_mandatory(library, mandatory_library_fields)
+            if 'name' not in item['state']:
+                raise ValueError("Invalid format: missing mandatory field %s", 'state.name')
+            if 'name' not in item['severity']:
+                raise ValueError("Invalid format: missing mandatory field %s", 'severity.name')
+
+            # Possible status as per checkmarx 9.2: TO_VERIFY, NOT_EXPLOITABLE, CONFIRMED, URGENT, PROPOSED_NOT_EXPLOITABLE
+            status = item['state']['name']
+            cve = item.get('cveName', 'NC')
+            finding_item = Finding(
+                title='{0} {1} | {2}'.format(library['name'], library['version'], cve),
+                severity=item['severity']['name'],
+                description=item.get('description', 'NC'),
+                unique_id_from_tool=item.get('id', None),
+                references=item.get('url', None),
+                mitigation=item.get('recommendations', 'No mitigation provided'),
+                impact='No impact provided',
+                component_name=library['name'],
+                component_version=library['version'],
+                cve=cve,
+                # 1035 is "Using Components with Known Vulnerabilities"
+                # Possible improvment: get the CWE from the CVE using some database?
+                # nvd.nist.gov has the info; see for eg https://nvd.nist.gov/vuln/detail/CVE-2020-25649 "Weakness Enumeration"
+                cwe=1035,
+                cvssv3_score=item.get('score', None),
+                publish_date=datetime.strptime(item['publishDate'], '%Y-%m-%dT%H:%M:%S'),
```

Comment:
are you sure this field is always there?

Reply:
nope; adding a verification...